rust/worker/src/execution/orchestration/compact.rs

```diff
@@ -237,6 +243,53 @@ impl CompactOrchestrator {
     ) {
         self.state = ExecutionState::Write;
 
+        let segments = self
+            .sysdb
+            .get_segments(
+                None,
+                Some(SegmentType::BlockfileMetadata.into()),
+                None,
+                Some(self.collection_id),
+            )
+            .await;
+
+        let segment = match segments {
+            Ok(mut segments) => {
+                if segments.is_empty() {
+                    return;
+                }
+                segments.drain(..).next().unwrap()
+            }
+            Err(_) => {
+                return;
+            }
+        };
+
+        if segment.r#type != SegmentType::BlockfileMetadata {
+            return;
+        }
+
+        tracing::debug!("Found metadata segment {:?}", segment);
+
+        let metadata_writer =
+            match MetadataSegmentWriter::from_segment(&segment, &self.blockfile_provider).await {
```

Comment:
can you remind me why we can't put this in get_segment_writers()?

Reply:
`get_segment_writers()` takes a mutable reference to self since it sets some of its fields. Metadatawriter has a lifetime since it contains a reader. If we create a metadatawriter inside `get_segment_writers()` then the compiler infers that the writer's lifetime is tied to the lifetime of self (since that is the only parameter to the method that is a reference). Thus we can't modify self in the method anymore because it is as if the writer is holding a reference to self.

If we shift the responsibility of modifying to the callsite instead then we are just shifting the problem one level down since now in the callsite, we'll have a similar problem since the writer has an immutable reference to self and is in scope at the callsite. If we inline it then there is no such implicit association between self and the writer

Reply:
With 'static, I am able to create the writer inside `get_segment_writers()` so made that change